internal/core/metadata/operators/device_profile/value_descriptor.go

```diff
@@ -0,0 +1,196 @@
+package device_profile
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/data/errors"
+)
+
+// ValueDescriptorAdder provides the necessary functionality for creating a ValueDescriptor.
+type ValueDescriptorAdder interface {
+	Add(vdr *contract.ValueDescriptor, ctx context.Context) (string, error)
+}
+
+// ValueDescriptorAdder provides the necessary functionality for updating a ValueDescriptor.
+type ValueDescriptorUpdater interface {
+	ValueDescriptorsUsage(names []string, ctx context.Context) (map[string]bool, error)
+	Add(vdr *contract.ValueDescriptor, ctx context.Context) (string, error)
+	Update(vdr *contract.ValueDescriptor, ctx context.Context) error
+	DeleteByName(name string, ctx context.Context) error
+	ValueDescriptorForName(name string, ctx context.Context) (contract.ValueDescriptor, error)
+}
+
+// ValueDescriptorAddExecutor creates ValueDescriptor(s) via the operator pattern.
+type ValueDescriptorAddExecutor interface {
+	Execute() error
+}
+
+type addValueDescriptor struct {
+	ctx    context.Context
+	drs    []contract.DeviceResource
+	client ValueDescriptorAdder
+	logger logger.LoggingClient
+}
+
+// Execute creates the necessary ValueDescriptors for a set of DeviceResources.
+func (a addValueDescriptor) Execute() error {
+	for _, dr := range a.drs {
+
+		desc := contract.From(dr)
+		id, err := a.client.Add(&desc, a.ctx)
+		if err != nil {
+			a.logger.Error(fmt.Sprintf("Unable to create value descriptor: %s", err.Error()))
+			return err
+		}
+		a.logger.Debug(fmt.Sprintf("Created Value Descriptor id: %s", id))
+	}
+	return nil
+}
+
+// NewAddExecutor creates a new ValueDescriptorAddExecutor.
+func NewAddExecutor(ctx context.Context, client ValueDescriptorAdder, loggingClient logger.LoggingClient, drs ...contract.DeviceResource) ValueDescriptorAddExecutor {
+	return addValueDescriptor{
+		ctx:    ctx,
+		drs:    drs,
+		client: client,
+		logger: loggingClient,
+	}
+}
+
+// updateValueDescriptor encapsulates the data needed to update a value descriptor.
+type updateValueDescriptor struct {
+	ctx    context.Context
+	dp     contract.DeviceProfile
+	loader DeviceProfileLoader
+	client ValueDescriptorUpdater
+	logger logger.LoggingClient
+}
+
+// UpdateValueDescriptorExecutor updates a value descriptor.
+type UpdateValueDescriptorExecutor interface {
+	Execute() error
+}
+
+// Execute updates a value descriptor with the provided information.
+func (u updateValueDescriptor) Execute() error {
```

Comment:
Normally when I see something simple require a lot of leg work in a micro-service architecture, like I am doing here, I look to see if we should either relocate a responsibility or create a new micro-service to handle it. I do not know if either of these are good solutions, but I am not the biggest fan of doing all this work for a simple update. This gets even more complicated if we try to ensure the state of the system during an error, like if we update ValueDescriptors and then fail to update the DeviceProfile. I feel like we cannot handle that situation elegantly with the current layout. Just a thought a had while doing this work.